cpp/src/arrow/columnar_format.cc

```diff
@@ -0,0 +1,1020 @@
+  for (int i = 0; i < static_cast<int>(columns_.size()); i++) {
+    if (columns_[i].field->Equals(*field)) {
+      return i;
+    }
+  }
+
+  return -low-1;
+}
+
+ColumnMap::ColumnMap() : impl_(new Impl()) {
+}
+
+ColumnMap::~ColumnMap() {
+}
+
+void ColumnMap::Put(const std::shared_ptr<Field>& field,
+                    const std::shared_ptr<Int16Array>& rep_levels,
+                    const std::shared_ptr<Int16Array>& def_levels,
+                    const std::shared_ptr<Array>& values) {
+  int i = impl_->find(field);
+  if (i < 0) {
+    i = -i-1;
+    impl_->columns_.insert(impl_->columns_.begin() + i, Impl::ColumnData());
+    impl_->columns_[i].field = field;
+  }
+  impl_->columns_[i].rep_levels = rep_levels;
+  impl_->columns_[i].def_levels = def_levels;
+  impl_->columns_[i].values = values;
+}
+
+int ColumnMap::find(const std::shared_ptr<Field>& field) const {
+  int i = impl_->find(field);
+  return (i >= 0) ? i : -1;
+}
+
+int ColumnMap::size() const {
+  return static_cast<int>(impl_->columns_.size());
+}
+
+void ColumnMap::get(int i, std::shared_ptr<Int16Array>* rep_levels,
+                           std::shared_ptr<Int16Array>* def_levels,
+                           std::shared_ptr<Array>* values) const {
+  if (rep_levels) {
+    *rep_levels = impl_->columns_[i].rep_levels;
+  }
+  if (def_levels) {
+    *def_levels = impl_->columns_[i].def_levels;
+  }
+  if (values) {
+    *values = impl_->columns_[i].values;
+  }
+}
+
+const std::shared_ptr<Int16Array>& ColumnMap::rep_levels(int i) const {
+  return impl_->columns_[i].rep_levels;
+}
+
+const std::shared_ptr<Int16Array>& ColumnMap::def_levels(int i) const {
+  return impl_->columns_[i].def_levels;
+}
+
+const std::shared_ptr<Array>& ColumnMap::values(int i) const {
+  return impl_->columns_[i].values;
+}
+
+
+// Utility functions and classes
+namespace {
+
+template<typename data_type>
+struct supports_basic_append {
+  static constexpr bool value =
+    has_c_type<data_type>::value ||
+    std::is_same<data_type, BooleanType>::value ||
+    std::is_base_of<FixedSizeBinaryType, data_type>::value;
+};
+
+/// Copy single value from array to builder, generic version
+template<typename data_type>
+typename std::enable_if<supports_basic_append<data_type>::value, Status>::type
+CopyValue(const typename TypeTraits<data_type>::ArrayType& array, int64_t i,
+          typename TypeTraits<data_type>::BuilderType* builder) {
+  return builder->Append(array.Value(i));
+}
+
+
+/// Copy single value from array to builder, specialization for BinaryType
+template<typename data_type>
+typename std::enable_if<std::is_base_of<BinaryType, data_type>::value, Status>::type
+CopyValue(const typename TypeTraits<data_type>::ArrayType& array, int64_t i,
+          typename TypeTraits<data_type>::BuilderType* builder) {
+  int32_t length;
+  const uint8_t* value = array.GetValue(i, &length);
+  return builder->Append(value, length);
+}
+
+
+#define PRIMITIVE_CASE_LIST \
```

Comment:
it might be clearer if you make CASE a parameter here.

Reply:
agreed, then below it could be used as:
```c++
switch (field->type()->id()) {
  #define CASE(data_type) \
    case data_type::type_id: \
      return PrimitiveStitchNode<data_type>::Create(field, rep_level, def_level, pool, node);
    PRIMITIVE_CASE_LIST(CASE)
  #undef CASE
```